### ISSUE TITLE
Unit tests for byte code instrumentation with test source files / methods

### ISSUE THREAD

**Issue by eirbjo** [WinterTechForum/skybar#15]
Comment from @marshallpierce:

How feasible is it to build some automated tests for this? Could we have some classes that exhibit various structures (loops, no loops, throws in the middle, doesn't throw, etc) and do classloader / asm wizardry to load the visited versions of them, run them, and verify the results?

**Comment by eirbjo** [comment]
I have implemented a small test showing how this could work. See SkybarClassVisitorTest and the instrumentClass. Example source files are in the testcases package. Have only added oneLiner tests so far. Code is not pretty, wanted to show it can work first. Thoughts, @marshallpierce ?

**Comment by marshallpierce** [comment]
Looks great. I'll add on more thorough tests when I get a chance.

**Comment by marshallpierce** [comment]
I made https://github.com/WinterTechForum/skybar/tree/feature/instrumentation-tests to start exploring this. There are several tests that fail with things like curly braces that end up with a 1 execution count, but `shouldCountForLoopWithException` looks particularly fishy.

**Comment by eirbjo** [comment]
I fixed shouldCountStaticInitOneLiner. Since no method was executed on the class, the class was never initialised and the <cinit> was never called. I rewrote class loading to use Class.forName since that allows an initialize parameter to be set to true.

I also needed to add an assert on the implicit return javac adds on line 8

**Comment by eirbjo** [comment]
The for-loop in ForLoopWithException was counted twice. Turns out the i=0 and the i++ makes two distinct LINENUMBER byte codes. Since we iterate over the LINENUMBERs to make local variables, line 9 had two local variables for the same line. So the count was double the expected. I've fixed this by making local variables for the unique set of line numbers instead.

**Comment by eirbjo** [comment]
shouldCountTryWithResources: Javac injects the close (with null checking and exception handling) on line 12. Try it in a debugger, you should see it break on line 12 exactly once.

**Comment by eirbjo** [comment]
SkybarClassVisitor test now runs green. I fixed some cases where javac inserts lines where there is no statements in Java source form. For constructors without explicit super() calls, javac will inject an invoke for that and put it on the line of the constructor declaration (or class declaration in case of an anonymous instance initialiser). Also, for implicit returns, javac will generate a return on the line number of the closing curly brace of the method body.

**Comment by eirbjo** [comment]
Really great that you found the double report of throw-exits and the double report for for-loops. That would take a while to find and figure out without automated tests like these.

Do you think this is ready for merging into master now?

**Comment by eirbjo** [comment]
@marshallpierce I've rewritten tests to parse the expected counts for each line out of the source code itself. The idea is that this should make tests easier to read and a lot easier to modify. Thoughts?

**Comment by marshallpierce** [comment]
This looks great. Your investigation is in line with my vague suspicions, but you got there faster and with less guessing than I would have. :) I really must learn more bytecode... I like the in-source execution counts too.

**Comment by eirbjo** [comment]
Here's a few tricks that removed some of the guesswork for me:

1) I added trace printing of original byte code and modified byte code to stdout. This way we can see what gets instrumented
2) I debugged by setting breakpoints on the bootstrap method(s). This helped me realise that registration was done twice for the same line number. And also that reporting was done for both the throw and the catch.
3) It always helps to have authored the buggy code in the first place :-)
